### ISSUE TITLE
Auto Login workaround for Windows Auth in the absence of eventing

### ISSUE THREAD

**Issue by lenniebriscoe** [IdentityServer/IdentityServer4#747]
From the issues list I can see that eventing has yet to be completed for IdentityServer4. I would like to implement auto login for windows authenticated users for a .net 4.6 client connecting to the Idp. A [previous issue](https://github.com/IdentityServer/IdentityServer3/issues/2807) @leastprivilege mentioned that this could be done using the PreAuthenticateAsync event and checking IP addresses to set the Idp on SignInMessage. Is there a work around for this in the meantime?

**Comment by leastprivilege** [comment]
Not sure how eventing is related to that. Also PreAuthenticateAsync does not exist anymore.

But there are multiple ways to achieve this - e.g.

* disable local logins and configure Windows to be the only ext provider
* hard code the Windows challenge in the login controller

**Comment by lenniebriscoe** [comment]
Probably my misunderstanding of what @brockallen meant by ['Eventing'](https://github.com/IdentityServer/IdentityServer4/issues/540) . Not to worry.

We want to allow internal windows users to access the site seamlessly, whereas external users to be prompted for username and password if not authenticated. Therefore it seems hardcoding the windows challenge in the login controller is the way forward. 

I've set up the IdentityServer4 so its running under IISExpress with both Windows and Anonymous Auth enabled, however when the login method gets hit I have an empty ClaimsIdentity rather than a Windows Identity. The Quickstart(8) code therefore doesnt have the User.Identity.Name which is needed to call .SignInAsync. How should I set this up so that I can have both Windows and Anonymous users authenticated ? Apparently it is possible (as you mentioned [here](https://github.com/IdentityServer/IdentityServer3/issues/1232))

**Comment by leastprivilege** [comment]
Check our quickstart UI repo

**Comment by lenniebriscoe** [comment]
I have been through the code and the only difference was I was missing the SecurityHeaders Attribute in my .net 4.6.1 mvc client and not using the hosted IdentityServer on demo.identityserver.io. I can't see how to configure the IdentityServer4 from that repo, nor is there any difference in the ExternalLogin action in the account controller. My older framework client presumeably only supports One of Anonymous OR Windows Authentication (I had both ticked in project properties but unsure?), but the IdentityServer4 Server using IISExpress and Kestrel should be able to support both?

Unless I'm missing something fundamental (more than likely!) I can't see anything that's helping?

**Comment by lenniebriscoe** [comment]
OK. so i have it working with mixed mode. I think I need a better way of determining if someone is internal to the organisation but this will do for now. Thanks for the assistance. Think it was the missing SecurityHeadersAttribute in my client.

**.net 4.6.1 client hosted in IISExpress**
All Owin related nuget packages referenced (inc Cookies\OpenIdConnect)
Startup.cs with Configuration method using Cookie and OpenIdConnect Authentication
In VS IDE, ensure the Anonymous and Windows Authentication are Enabled on the website properties
SecurityHeaders and Authorize attribute on Controllers\Actions
Implicit grant type

**IdentityServer4 .net core server hosted in IISExpress using Kestrel**
In VS IDE, ensure the Anonymous and Windows Authentication are Enabled on the project properties
SecurityHeaders attribute on Controllers
In Account Controller HttpGet Login Action checks hostname is a local host and then directly calls ExternalLogin("Negotiate", returnUrl) otherwise continue as normal.

User that has an internal host goes straight in
User that has an external host gets redirected to the login page

**Comment by leastprivilege** [comment]
cool!

**Comment by BootsThunderclap** [comment]
I used a very similar technique, but @lenniebriscoe's solution was my jumping off point. I used DI to inject an instance IHostingEnvironment into the AccountController. Basically, I wanted developers and testers to be able to log in as local test users to be able to test various authorization policies. If the environment is prod or UAT, I call the ExternalLogin action directly. If it's a test or development environment, I just let it continue with the normal flow. This seems to simulate a single-sign-on experience pretty well.
